syz-verifier/exectask.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2021 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package main
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/google/syzkaller/pkg/rpctype"
+	"github.com/google/syzkaller/prog"
+)
+
+type EnvDescr int64
+
+const (
+	AnyEnvironment EnvDescr = iota
+	NewEnvironment
+	// TODO: add CleanVMEnvironment support.
+
+	EnvironmentsCount
+)
+
+// ExecTask is the atomic analysis entity. Once executed, it could trigger the
+// pipeline propagation fof the program.
+type ExecTask struct {
+	CreationTime   time.Time
+	Program        *prog.Prog
+	ID             int64
+	ExecResultChan ExecResultChan
+
+	priority int // The priority of the item in the queue.
+	// The index is needed by update and is maintained by the heap.Interface methods.
+	index int // The index of the item in the heap.
+}
+
+func (t *ExecTask) ToRPC() *rpctype.ExecTask {
+	return &rpctype.ExecTask{
+		Prog: t.Program.Serialize(),
+		ID:   t.ID,
+	}
+}
+
+var (
+	ChanMapMutex           = sync.Mutex{}
+	TaskIDToExecResultChan = map[int64]ExecResultChan{}
+	TaskCounter            = int64(-1)
+)
+
+type ExecResultChan chan *ExecResult
+
+func MakeExecTask(prog *prog.Prog) *ExecTask {
+	task := &ExecTask{
+		CreationTime:   time.Now(),
+		Program:        prog,
+		ExecResultChan: make(ExecResultChan),
+		ID:             atomic.AddInt64(&TaskCounter, 1),
+	}
+
+	ChanMapMutex.Lock()
+	defer ChanMapMutex.Unlock()
+	TaskIDToExecResultChan[task.ID] = task.ExecResultChan
+
+	return task
+}
+
+func DeleteExecTask(task *ExecTask) {
+	ChanMapMutex.Lock()
+	defer ChanMapMutex.Unlock()
+	delete(TaskIDToExecResultChan, task.ID)
+}
+
+func GetExecResultChan(taskID int64) ExecResultChan {
+	ChanMapMutex.Lock()
+	defer ChanMapMutex.Unlock()
+
+	return TaskIDToExecResultChan[taskID]
+}
+
+func MakeExecTaskQueue() *ExecTaskQueue {
+	return &ExecTaskQueue{
+		pq: make(ExecTaskPriorityQueue, 0),
+	}
+}
+
+// ExecTaskQueue respects the pq.priority. Internally it is a thread-safe PQ.
+type ExecTaskQueue struct {
+	pq ExecTaskPriorityQueue
+}
+
+// PopTask return false if no tasks are available.
+func (q *ExecTaskQueue) PopTask() (*ExecTask, bool) {
+	if q.pq.Len() == 0 {
+		return nil, false
+	}
+
+	return q.pq.Pop().(*ExecTask), true
+}
+
+func (q *ExecTaskQueue) PushTask(task *ExecTask) {
+	q.pq.Push(task)
+}
+
+func (q *ExecTaskQueue) Len() int {
+	return q.pq.Len()
+}
+
+// ExecTaskPriorityQueue reused example from https://pkg.go.dev/container/heap
```

Comment:
I am not following. This is not really using any of the heap package functions. This does not even import "container/heap".

Reply:
fixed